README.md
# ACME-Cli

[![Build Status](https://travis-ci.org/zealot128/ruby-acme-cli.svg?branch=travis)](https://travis-ci.org/zealot128/ruby-acme-cli)
[![Gem Version](https://badge.fury.io/rb/acme-cli.svg)](https://badge.fury.io/rb/acme-cli)

Yet another ACME client (e.g. to use together with Letsencrypt CA to issue TLS certs) for command lines using Ruby.

## Installation

* This tool needs Ruby >= 2.1 (as the dependency ``acme-client`` needs that because of use of keyword arguments).
* OpenSSL bindings
* no sudo! (needs access to webserver-root ``/.well-known/acme-challenges`` alias for all domains - See later section for Nginx example)

```
# check your ruby version:
$ ruby --version
ruby 2.2.3p173 (2015-08-18 revision 51636) [x86_64-linux]

$ gem install acme-cli

$ acme-cli --version
0.2.0
```

### Troubleshooting Ruby version

Unfortunately, most Linux distributions does not ship a current Ruby version (Version 1.9.3 or 2.0). Check, if your ruby version is at least 2.2. Otherwise you need to update the Ruby.

If you are installing this as a non-root user, you might want to try RVM. Installation itself needs no root, but needs some packages:

```
sudo apt-get install curl bison build-essential zlib1g-dev libssl-dev libreadline6-dev libxml2-dev libgmp-dev git-core
```

To install RVM:

```
gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3
\curl -sSL https://get.rvm.io | bash -s stable  --autolibs=disable --auto-dotfiles

rvm install 2.2
source ~/.bashrc # or ~/.profile RVM tells you to reload your shell

ruby --version
```

Notice: If you are using RVM, all your cronjobs must be run as a login shell, otherwise RVM does not work:

```cron
* * * * * /bin/bash -l -c "acme-cli manage ..."
```

Another way, e.g. on Ubuntu 14.04 might be to use the [Brightbox ppa](https://www.brightbox.com/blog/2015/01/05/ruby-2-2-0-packages-for-ubuntu/).

## Usage

Specify ``-t`` to use Letsencrypt test server. Without it, all requests are called against the production server, that might have some more strict rate limiting. If you are just toying around, add the -t flag.

```bash
# show all commands

acme-cli help

# show options for an individual command
acme-cli help cert

# creates account_key.json in current_dir
acme-cli register -t myemail@example.com

# authorize one or more domains/subdomains
acme-cli authorize -t --webroot-path /var/www/default example.com www.example.com somedir.example.com

# experimental: authorize all server_names in /etc/nginx/sites-enabled/*
acme-cli authorize_all -t --webroot-path /var/www/default

# create a certificate for domains that are already authorized within the last minutes (1h-2h I think)
# the first domain will be the cn subject. All other are subjectAlternateName
# if cert.pem already exists, will only create a new one if the old is expired
# (30 days before expiration) -> see full help
acme-cli help cert

acme-cli cert -t example.com www.example.com somdir.example.com
# will create key.pem fullchain.pem chain.pem and cert.pem in current directory

# checks validation date of given certificate.
# Exists non-zero if:
# * not exists (exit 1)
# * will expire in more than 30 days (exit code 2)
acme-cli check --days-valid 30 cert.pem
```


And last but not least, the meta command ``manage`` that integrated check + authorize + cert (intended to be run as cronjob):

```bash
$ acme-cli manage --days-valid 30 \
                       --account-key /home/acme/account_key.pem \
                       --webroot-path /home/acme/webroot/.well-known/acme-challenge \
                       --key-directory /home/acme/certs \
                       example.com www.example.com

2015-12-05 23:40:04 +0100: Certificate /home/acme/certs/example.com/cert.pem does not exists
2015-12-05 23:40:04 +0100: Authorizing example.com...
2015-12-05 23:40:04 +0100: existing account key found
2015-12-05 23:40:06 +0100: Authorization successful for example.com
2015-12-05 23:40:06 +0100: Authorizing www.example.com
2015-12-05 23:40:08 +0100: Authorization successful for www.example.com
2015-12-05 23:40:08 +0100: creating new private key to /home/acme/certs/example.com/key.pem...
2015-12-05 23:40:09 +0100: Certificate successfully created to /home/acme/certs/example.com/fullchain.pem /home/acme/certs/example.com/chain.pem and /home/acme/certs/example.com/cert.pem!
2015-12-05 23:40:09 +0100: Certificate valid until: 2016-03-04 21:40:00 UTC

# Run command again exits immediately:
$ acme-cli manage --days-valid 30 --account-key /home/acme/account_key.pem --webroot-path /home/acme/webroot/.wel
l-known/acme-challenge --key-directory /home/acme/certs \
      example.com www.example.com
2015-12-05 23:40:17 +0100: Certificate '/home/acme/certs/example.com/cert.pem' valid until 2016-03-04.
$ echo $?
1
```

This had:

1. check if /home/acme/certs/example.com/cert.pem exists and expires in less than 30 days (or exit 1 at this point)
2. authorize all domains + subdomains
3. issue one certificate with those domains and place it under /home/acme/certs/example.com/[key.pem,fullchain.pem,chain.pem,cert.pem]
4. exit 0 -> so can be && with ``service nginx reload`` or mail deliver

For running as cron, reducing log level to fatal might be desirable: ``acme-cli manage --log-level fatal``.

## Example integration Nginx:

```nginx
server {
  listen 80;
  server_name example.com www.example.com somedir.example.com
  location /.well-known/acme-challenge {
	  alias /home/acme/webroot/.well-known/acme-challenge;
	  default_type "text/plain";
	  try_files $uri =404;
  }
```

notice the location - alias. Use this dir with ``--webroot-path`` for authorization.

Afterwards, use the fullchain.pem and key.pem:

```nginx
server {
  listen 443 ssl;
  server_name example.com www.example.com;
  ssl on;
  ssl_certificate_key /home/acme/certs/example.com/key.pem;
  ssl_certificate /home/acme/certs/example.com/fullchain.pem;

  # use the settings from: https://gist.github.com/konklone/6532544
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release` to create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

1. Fork it ( https://github.com/zealot128/ruby-acme-cli/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# Change Log

## [v0.6.0](https://github.com/zealot128/ruby-acme-cli/tree/v0.6.0)

* Rename to acme-client for trademark reasons (#22)

## [v0.5.0](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.5.0)

* fix for CSR generated using a pre-1.0.2 OpenSSL with a client that doesn't properly specify the CSR version. See https://community.letsencrypt.org/t/openssl-bug-information/19591 (Acme::Client::Error::Malformed) https://github.com/zealot128/ruby-letsencrypt-cli/commit/b7fd1d592e9a74905f5067b64e0ac88a526cfeed
* explicitly require colorize Gem so color support does work https://github.com/zealot128/ruby-letsencrypt-cli/commit/b43510d1be1a495923ea8e27051b3a0bae4e23b0

## [v0.4.1](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.4.1)

*  fix renewing via manage command when certificate is not expired https://github.com/zealot128/ruby-letsencrypt-cli/commit/8e6b9cd4a2b1d0caa7a85d6ead410b98555cb499
*  require logger in beginning of file

## [v0.4.0](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.4.0)

* New ``--sub-directory`` option to use that instead of first domain name
* Solves issue #10 -- certificate_exists_and_valid_and_all_domains_included? returns true when cert is expired

## [v0.3.0](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.3.0)

* Certificate creation checks if existing certificate includes all requested domains. If at least one is missing, a new cert will be requested
* Added Ruby 2.3.0 and Ruby head to the build matrix

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.2.0...v0.3.0)

## [v0.2.0](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.2.0)

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.4...v0.2.0)

**Closed issues:**

- cf1e0d9 Exit code 2, if certificate is still valid

**Merged pull requests:**

- Apply strict permissions on private key [\#4](https://github.com/zealot128/ruby-letsencrypt-cli/pull/4) ([zygiss](https://github.com/zygiss))
- Fix typo in README [\#2](https://github.com/zealot128/ruby-letsencrypt-cli/pull/2) ([kenrick](https://github.com/kenrick))

## [v0.1.4](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.4) (2015-12-08)

* require higher acme-client version, that generated correct fullchain certs.
  fullchain.pem is chain.pem + cert.pem, should be cert.pem + chain.pem [\#1](https://github.com/zealot128/ruby-letsencrypt-cli/issues/1)

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.3...v0.1.4)

## [v0.1.3](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.3) (2015-12-06)

* Fixed registration
* Added various specs

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.2...v0.1.3)

## [v0.1.2](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.2) (2015-12-05)

* Added manage command
* Improved nginx doc + Ruby installation

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.1...v0.1.2)

## [v0.1.1](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.1) (2015-12-05)

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.0...v0.1.1)

* b654469 new command: check PATH_TO_CERT

## [v0.1.0](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.0) (2015-12-05)

* released first public version
* added --version flag
* added explicit production server

[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.0.beta1...v0.1.0)

## [v0.1.0.beta1](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.0.beta1) (2015-12-05)
[Full Changelog](https://github.com/zealot128/ruby-letsencrypt-cli/compare/v0.1.0.pre...v0.1.0.beta1)

## [v0.1.0.pre](https://github.com/zealot128/ruby-letsencrypt-cli/tree/v0.1.0.pre) (2015-12-05)


\* *This Change Log was (partially) automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


spec/spec_helper.rb
require 'simplecov'
require 'timecop'
require 'vcr'
require 'webmock'
require 'pry'
require 'active_support/all'
SimpleCov.start
Dir['spec/support/**/*'].each {|f| load f }

VCR.configure do |config|
  config.cassette_library_dir = "spec/fixtures/vcr_cassettes"
  config.hook_into :webmock # or :fakeweb
end

RSpec.configure do |config|
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end
  config.backtrace_exclusion_patterns << %r{/gems/}
  config.disable_monkey_patching!

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/support/thor_helpers.rb

module ThorHelpers
  # http://apidock.com/rails/v4.2.1/Kernel/capture
  def capture(stream)
    stream = stream.to_s
    captured_stream = Tempfile.new(stream)
    stream_io = eval("$#{stream}")
    origin_stream = stream_io.dup
    stream_io.reopen(captured_stream)

    yield

    stream_io.rewind
    return captured_stream.read
  ensure
    captured_stream.close
    captured_stream.unlink
    stream_io.reopen(origin_stream)
  end
end

RSpec.configure do |config|
  config.around :each do |example|
    Dir.mktmpdir {|dir|
      @current_dir = dir
      example.run
    }
  end
  config.include ThorHelpers
end


lib/letsencrypt/cli/acme_wrapper.rb
require 'json'
require 'acme-client'
require 'logger'
require 'colorize'

class AcmeWrapper
  def initialize(options)
    @options = options
    if !@options[:color]
      String.disable_colorization = true
    end
  end

  def log(message, severity=:info)
    @logger ||= Logger.new(STDOUT).tap {|logger|
      logger.level = Logger::SEV_LABEL.index(@options[:log_level].upcase)
      logger.formatter = proc do |sev, datetime, progname, msg|
        "#{datetime.to_s.light_black}: #{msg}\n"
      end
    }
    @logger.send(severity, message)
  end

  def client
    @client ||= Acme::Client.new(private_key: account_key, endpoint: endpoint)
  end

  def authorize(domain)
    FileUtils.mkdir_p(@options[:webroot_path])
    log "Authorizing #{domain.blue}.."
    authorization = client.authorize(domain: domain)

    challenge = authorization.http01

    challenge_file = File.join(@options[:webroot_path], challenge.filename.split('/').last)
    log "Writing challenge to #{challenge_file}", :debug
    File.write(challenge_file, challenge.file_content)

    challenge.request_verification

    5.times do
      log "Checking verification...", :debug
      sleep 1
      break if challenge.verify_status != 'pending'
    end
    if challenge.verify_status == 'valid'
      log "Authorization successful for #{domain.green}"
      File.unlink(challenge_file)
      true
    else
      log "Authorization error for #{domain.red}", :error
      log challenge.error['detail']
      false
    end
  end

  def cert(domains)
    return if certificate_exists_and_valid_and_all_domains_included?(domains)
    csr = OpenSSL::X509::Request.new
    certificate_private_key = find_or_create_pkey(@options[:private_key_path], "private key", @options[:key_length] || 2048)

    csr.subject = OpenSSL::X509::Name.new([
      # ['C',             options[:country], OpenSSL::ASN1::PRINTABLESTRING],
      # ['ST',            options[:state],        OpenSSL::ASN1::PRINTABLESTRING],
      # ['L',             options[:city],         OpenSSL::ASN1::PRINTABLESTRING],
      # ['O',             options[:organization], OpenSSL::ASN1::UTF8STRING],
      # ['OU',            options[:department],   OpenSSL::ASN1::UTF8STRING],
      # ['CN',            options[:common_name],  OpenSSL::ASN1::UTF8STRING],
      # ['emailAddress',  options[:email],        OpenSSL::ASN1::UTF8STRING]
      ['CN', domains.first, OpenSSL::ASN1::UTF8STRING]
    ])
    if domains.count > 1
      ef = OpenSSL::X509::ExtensionFactory.new
      exts = [ ef.create_extension( "subjectAltName", domains.map{|domain| "DNS:#{domain}"}.join(','), false  ) ]
      attrval = OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence(exts)])
      attrs = [
        OpenSSL::X509::Attribute.new('extReq', attrval),
        OpenSSL::X509::Attribute.new('msExtReq', attrval),
      ]
      attrs.each do |attr|
        csr.add_attribute(attr)
      end
    end
    csr.version = 2
    csr.public_key = certificate_private_key.public_key
    csr.sign(certificate_private_key, OpenSSL::Digest::SHA256.new)
    certificate = client.new_certificate(csr)
    File.write(@options[:fullchain_path], certificate.fullchain_to_pem)
    File.write(@options[:chain_path], certificate.chain_to_pem)
    File.write(@options[:certificate_path], certificate.to_pem)
    log "Certificate successfully created to #{@options[:fullchain_path]} #{@options[:chain_path]} and #{@options[:certificate_path]}!".green
    log "Certificate valid until: #{certificate.x509.not_after}"
  end

  def check_certificate(path)
    unless File.exists?(path)
      log "Certificate #{path} does not exists", :warn
      return false
    end
    cert = OpenSSL::X509::Certificate.new(File.read(path))
    renew_on = cert.not_after.to_date - @options[:days_valid]
    log "Certificate '#{path}' valid until #{cert.not_after.to_date}.", :info
    if Date.today >= renew_on
      log "Certificate '#{path}' should be renewed!", :warn
      return false
    else
      true
    end
  end

  def revoke_certificate(path)
    unless File.exists?(path)
      log "Certificate #{path} does not exists", :warn
      return false
    end
    cert = OpenSSL::X509::Certificate.new(File.read(path))
    if client.revoke_certificate(cert)
      log "Certificate '#{path}' was revoked", :info
    end
    true
  rescue Acme::Client::Error::Malformed => e
    log e.message, :error
    return false
  end

  private

  def certificate_exists_and_valid_and_all_domains_included?(domains)
    return false if !File.exists?(@options[:certificate_path])
    cert = OpenSSL::X509::Certificate.new(File.read(@options[:certificate_path]))
    domains_in_cert = cert.extensions.map(&:to_h).select{|i| i['oid'] == 'subjectAltName' }.map{|i| i['value']}.join(', ').split(/, */).map{|i| i.sub(/^DNS:/, '') }  +
      [ cert.subject.to_s.sub(%r{/CN=}, '') ].uniq.sort
    missing_domains = domains.sort.uniq - domains_in_cert
    if missing_domains != []
      log "Certificate '#{@options[:certificate_path]}' missing domains #{missing_domains.join(' ')}. Existing: #{domains_in_cert.join(' ')}", :warn
      return false
    end
    expires_on = cert.not_after.to_date
    if expires_on <= Date.today
      log "Certificate '#{@options[:certificate_path]}' has expired on #{expires_on}.", :warn
      return false
    end
    renew_on = expires_on - @options[:days_valid]
    if renew_on > Date.today
      log "Certificate '#{@options[:certificate_path]}' still valid till #{cert.not_after.to_date}.", :warn
      log "Won't renew until #{renew_on} (#{@options[:days_valid]} days before)", :warn
      exit 2
    end

    false 
  end

  def endpoint
    if @options[:test]
      "https://acme-staging.api.letsencrypt.org"
    else
      "https://acme-v01.api.letsencrypt.org"
    end
  end

  def account_key
    @account_key ||= find_or_create_pkey(@options[:account_key], "account key", @options[:key_length] || 4096)
  end

  def find_or_create_pkey(file_path, name, length)
    if File.exists?(file_path)
      log "existing account key found"
      OpenSSL::PKey::RSA.new File.read file_path
    else
      log "creating new private key to #{file_path}..."
      private_key = OpenSSL::PKey::RSA.new(length)
      File.write(file_path, private_key.to_s)
      File.chmod(0400, file_path)
      private_key
    end
  end
end


lib/letsencrypt/cli/app.rb
require 'thor'
require 'colorize'
require 'fileutils'
module Letsencrypt
  module Cli
    class App < Thor
      class_option :account_key, desc: "Path to private key file (will be created if not exists)", aliases: "-a", default: 'account_key.pem'
      class_option :test, desc: "Use staging url of Letsencrypt instead of production server", aliases: "-t", type: :boolean
      class_option :log_level, desc: "Log Level (debug, info, warn, error, fatal)", default: "info"
      class_option :color, desc: "Disable colorize", default: true, type: :boolean

      desc 'register EMAIL', 'Register account'
      method_option :key_length, desc: "Length of generated private key", type: :numeric, default: 4096
      def register(email)
        if email.nil? || email == ""
          wrapper.log "no E-Mail specified!", :fatal
          exit 1
        end
        if !email[/.*@.*/]
          wrapper.log "not an email", :fatal
          exit 1
        end
        registration = wrapper.client.register(contact: "mailto:" + email)
        registration.agree_terms
        wrapper.log "Account created, Terms accepted"
      end

      desc 'authorize_all', "Verify all server_names in /etc/nginx/sites-enabled/* (needs read access)"
      method_option :webroot_path, desc: "Path to mapped .acme-challenge folder (no subdir)", aliases: '-w', required: true
      method_option :webserver_dir, desc: "Path to webserver configs", default: "/etc/nginx/sites-enabled"
      def authorize_all
        lines = Dir[ File.join(@options[:webserver_dir], "*")].map{|file| File.read(file).lines.grep(/^\s*server_name/) }.flatten
        domains = lines.flatten.map{|i| i.strip.split(/[; ]/).drop(1) }.flatten.reject{|i| i.length < 3 }.uniq
        authorize(*domains)
      end

      desc 'authorize [DOMAINS]', 'Authorize all domains'
      method_option :webroot_path, desc: "Path to mapped .well-known/acme-challenge folder (no subdirs will be created)", aliases: '-w', required: true
      def authorize(*domains)
        rc = 0
        domains.each do |domain|
          if !wrapper.authorize(domain)
            rc = 1
          end
        end
        if rc != 0
          exit rc
        end
      end

      desc "cert [DOMAINS]", "create certificate and private key pair for domains. The first domain is the main CN domain, the reset will be added as SAN. If the given certificate-path already exists, script will exit non-zero if the certificate is still valid until the given number of days before."
      method_option :private_key_path, desc: "Path to private key. Will be created if non existant", aliases: '-k', default: 'key.pem'
      method_option :key_length, desc: "Length of private key", default: 2048, type: :numeric
      method_option :fullchain_path, desc: "Path to fullchain certificate (Nginx) (will be overwritten if exists!)", aliases: '-f', default: 'fullchain.pem'
      method_option :certificate_path, desc: "Path to certificate (Apache)", aliases: '-c', default: 'cert.pem'
      method_option :chain_path, desc: "Path to chain (Apache)", aliases: '-n', default: 'chain.pem'
      method_option :days_valid, desc: "If the --certificate-path already exists, only create new stuff, if that certificate isn't valid for less than the given number of days", default: 30, type: :numeric
      def cert(*domains)
        if domains.length == 0
          wrapper.log "no domains given", :fatal
          exit 1
        end
        wrapper.cert(domains)
      end

      desc "check PATH_TO_CERTIFICATE", "checks, if a given certificate exists and is valid until DAYS_VALID"
      method_option :days_valid, desc: "If the --certificate-path already exists, only create new stuff, if that certificate isn't valid for less than the given number of days", default: 30, type: :numeric
      def check(path)
        if !wrapper.check_certificate(path)
          exit 1
        end
      end

      desc "revoke PATH_TO_CERTIFICATE", "revokes a given certificate"
      def revoke(path)
        if !wrapper.revoke_certificate(path)
          exit 1
        end
      end

      desc "manage DOMAINS", "meta command that will: check if cert already exists / still valid (exits zero if nothing todo, exits 2 if certificate is still valid) + authorize given domains + issue certificate for given domains"
      method_option :key_length, desc: "Length of private key", default: 2048, type: :numeric
      method_option :days_valid, desc: "If the --certificate-path already exists, only create new stuff, if that certificate isn't valid for less than the given number of days", default: 30, type: :numeric
      method_option :webroot_path, desc: "Path to mapped .well-known/acme-challenge folder (no subdirs will be created)", aliases: '-w', required: true
      method_option :key_directory, desc: "Base directory for certificate storage.", default: "~/certs/"
      method_option :sub_directory, desc: "Sub-directory name in base directory where all certs + key are stored. If not set, the first domain name will be used", default: nil
      def manage(*domains)
        key_dir = File.join(@options[:key_directory], @options[:sub_directory] || domains.first)
        FileUtils.mkdir_p(key_dir)
        @options = @options.merge(
          :private_key_path  => File.join(key_dir, 'key.pem'),
          :fullchain_path    => File.join(key_dir, 'fullchain.pem'),
          :certificate_path  => File.join(key_dir, 'cert.pem'),
          :chain_path        => File.join(key_dir, 'chain.pem'),
        )
        if wrapper.check_certificate(@options[:certificate_path])
          exit 2
        end
        authorize(*domains)
        cert(*domains)
      end

      map %w[--version -v] => :__print_version
      desc "--version, -v", "print the version"
      def __print_version
        puts Letsencrypt::Cli::VERSION
      end

      private

      def wrapper
        @wrapper ||= AcmeWrapper.new(options)
      end
    end
  end
end
